src/server/common/dbconn.cpp

```diff
@@ -109,6 +110,61 @@ std::string dbconn::get_tournaments()
 	}
 }
 
+std::unique_ptr<simple_wml::document> dbconn::get_game_history(int player_id, int offset)
+{
+	try
+	{
+		std::string game_history_query = "select "
+"  game.GAME_NAME, "
+"  game.RELOAD, "
+"  game.START_TIME, "
+"  GROUP_CONCAT(CONCAT(player.USER_NAME, ':', player.FACTION)) as PLAYERS, "
+"  IFNULL(scenario.NAME, scenario.ID) as SCENARIO, "
+"  IFNULL(era.NAME, era.ID) as ERA, "
+"  IFNULL(GROUP_CONCAT(distinct mods.NAME), IFNULL(GROUP_CONCAT(distinct mods.ID), '')) as MODIFICATIONS, "
+"  concat('https://replays.wesnoth.org/', substring(game.INSTANCE_VERSION, 1, 4), '/', year(game.END_TIME), '/', lpad(month(game.END_TIME), 2, '0'), '/', lpad(day(game.END_TIME), 2, '0'), '/', game.REPLAY_NAME) as REPLAY_URL "
+"from wesnothd_game_info game "
+"inner join wesnothd_game_player_info player "
+"   on exists "
+"  ( "
+"    select 1 "
+"    from wesnothd_game_player_info player1 "
+"    where game.INSTANCE_UUID = player1.INSTANCE_UUID "
+"      and game.GAME_ID = player1.GAME_ID "
+"      and player1.USER_ID = ? "
+"  ) "
+"  and game.INSTANCE_UUID = player.INSTANCE_UUID "
+"  and game.GAME_ID = player.GAME_ID "
+"  and player.USER_ID != -1 "
+"inner join wesnothd_game_content_info scenario "
+"   on scenario.TYPE = 'scenario' "
+"  and scenario.INSTANCE_UUID = game.INSTANCE_UUID "
+"  and scenario.GAME_ID = game.GAME_ID "
+"inner join wesnothd_game_content_info era "
+"   on era.TYPE = 'era' "
+"  and era.INSTANCE_UUID = game.INSTANCE_UUID "
+"  and era.GAME_ID = game.GAME_ID "
+"left join wesnothd_game_content_info mods "
+"   on mods.TYPE = 'modification' "
+"  and mods.INSTANCE_UUID = game.INSTANCE_UUID "
+"  and mods.GAME_ID = game.GAME_ID "
+"group by game.INSTANCE_UUID, game.GAME_ID "
+"order by game.START_TIME desc "
+"limit 10 offset ? ";
```

Comment:
There appear to be performance considerations with high offsets so we should be careful to make sure we have no such issues or limit the offset amount.

Reply:
The potential performance issue as I understand it is that it can't just jump straight to the offset, rather it needs to also find everything else before that first.  Putting the index on the START_TIME column used in the `order by` should be enough, I think.  By the time we're in the range where even an indexed column has performance issues we'd be beyond the number of games any single person would've been able to play.

Reply:
True, I assume just using a high offset without a high number of games being in the DB should not be an issue and as you say we're not likely to get such a high games count.